man/res_file.5
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "RES_FILE 1"
.TH RES_FILE 1 "2014-09-13" "Clockwork v2.3.0" "Clockwork Manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
res_file \- Clockwork Resource Type for managing files
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
A fair portion of Linux system administration involves creating and
maintaining configuration files.  The \fBres_file\fR resource for
Clockwork automates much of this work by allowing an administrator
to specify almost every attribute of a file in a system policy:
.IP "1. Ownership (user and group)" 4
.IX Item "1. Ownership (user and group)"
.PD 0
.IP "2. Permissions" 4
.IX Item "2. Permissions"
.IP "3. Contents (both static and dynamic)" 4
.IX Item "3. Contents (both static and dynamic)"
.PD
.SH "ATTRIBUTES"
.IX Header "ATTRIBUTES"
.IP "\fBowner\fR" 4
.IX Item "owner"
The name of the user who should own the file.
.IP "\fBgroup\fR" 4
.IX Item "group"
The name of the group who should own this file.
.IP "\fBmode\fR" 4
.IX Item "mode"
File access permissions.  The value must be
specified in octal notation, i.e. 0755 for rwxr-xr-x.
.IP "\fBsource\fR" 4
.IX Item "source"
The path (from the policy master's point of view) to the source file.
If specified, the client-side file will be updated to match this static
file.
.Sp
Has no affect on directories, hard links or symbolic links.
.IP "\fBtemplate\fR" 4
.IX Item "template"
The path (from the policy master's point of view) to a source file template.
Like the \fBsource\fR
attribute, specifying this will cause the client-side file to be updated
from the copy on the master.  A template, however, is evaluated against
the client's set of facts and a custom source file is generated.
.Sp
Has no affect on directories, hard links or symbolic links.
.IP "\fBpath\fR" 4
.IX Item "path"
The path (from the client's point of view) to the file to be managed.
.IP "\fBpresent\fR" 4
.IX Item "present"
Whether or not this file should exist.  Valid values are
\&\*(L"yes\*(R" and \*(L"no\*(R".  The default value is \*(L"yes\*(R".
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.SS "Ownership and Permissions"
.IX Subsection "Ownership and Permissions"
Ensure ownership and permissions on /etc/sudoers:
.PP
.Vb 5
\&    file "/etc/sudoers" {
\&        owner: "root"
\&        group: "root"
\&        mode:  "0440" # required mode for sudo to work
\&    }
.Ve
.SS "Complete Control"
.IX Subsection "Complete Control"
In this scenario, local changes to /etc/foo.conf, including edits,
file removal, and ownership / permissions changes will be undone
by the Clockwork agent.
.PP
.Vb 6
\&    file "/etc/foo.conf" {
\&        owner:  "foo"
\&        group:  "root"
\&        mode:   "0640"
\&        source: "/clockwork/files/foo.conf"
\&    }
.Ve
.SS "Templates"
.IX Subsection "Templates"
Templates can be used to flexibly push out files that are similar
on all hosts, but differ with host-specific details.  Assuming the
following template (/clockwork/templates/banner.tpl):
.PP
.Vb 1
\&    <%= sys.fqdn %> \- Widgets Co. Ltd.
\&
\&    Unauthorized access to this machine is prohibited.
\&    Use of this system is limited to authorized individuals only.
\&    All activity is monitored.
.Ve
.PP
This host resource will enforce a banner for all hosts that contains
the standard \*(L"warning\*(R" verbiage and the local hostname:
.PP
.Vb 6
\&    file "/etc/banner" {
\&        owner:    "root"
\&        group:    "root"
\&        mode:     "0400"
\&        template: "/clockwork/templates/banner.tpl"
\&    }
.Ve
.PP
When evaluated on the host named 'labs.example.net', /etc/banner
will contain the following:
.PP
.Vb 1
\&    labs.example.net \- Widgets Co. Ltd.
\&
\&    Unauthorized access to this machine is prohibited.
\&    Use of this system is limited to authorized individuals only.
\&    All activity is monitored.
.Ve
.SH "CAVEATS"
.IX Header "CAVEATS"
.SS "1. Link Support"
.IX Subsection "1. Link Support"
Symbolic links and hard links cannot be created with the file
resource.  Use \fBres_symlink\fR(5) instead.
.SH "DEPENDENCIES"
.IX Header "DEPENDENCIES"
File resources implicitly create the following dependencies:
.SS "User Owner (owner)"
.IX Subsection "User Owner (owner)"
If the \fBowner\fR
attribute is specified, the file resource will depend on the
existence of that user.  This is designed to ensure that files are
owned by valid system users.
.SS "Group Owner (group)"
.IX Subsection "Group Owner (group)"
If the \fBgroup\fR
attribute is specified, the file resource will depend on the
existence of that group.  This is designed to ensure that files are
owned by valid system groups.
.SS "Parent Directories"
.IX Subsection "Parent Directories"
All parent components of the file path will be created if the file
should exist and does not.  By default, these parent directories
will be owned by root:root with mode 0755.  However, if the policy
defines \fBres_dir\fR(5) resources with paths matching any component
of the parent directory chain, the file resource will be set to
depend on them.
.PP
An example should clear up any confusion:
.PP
.Vb 6
\&    file "/u/apps/test.example.net/index.html" {
\&        owner: "apache"
\&        group: "web"
\&        mode:  0664
\&        source: "/srv/clockwork/www/under\-constr.html"
\&    }
\&
\&    dir "/u/apps" {
\&        owner: "webmaster"
\&        group: "web"
\&        mode:  0755
\&    }
\&
\&    dir "/u/apps/test.example.net" {
\&        owner: "apache"
\&        group: "web"
\&        mode:  0775
\&    }
.Ve
.PP
If the above policy was enforced on a host without the /u directory,
the following would happen (in order):
.IP "1. Create /u, per defaults (root:root; 0755)" 4
.IX Item "1. Create /u, per defaults (root:root; 0755)"
.PD 0
.IP "2. Create /u/apps, per policy (webmaster:web; 0755)" 4
.IX Item "2. Create /u/apps, per policy (webmaster:web; 0755)"
.IP "3. Create /u/apps/test.example.net, per policy (apache:web; 0775)" 4
.IX Item "3. Create /u/apps/test.example.net, per policy (apache:web; 0775)"
.IP "4. Create the file and retrieve its contents from the policy master." 4
.IX Item "4. Create the file and retrieve its contents from the policy master."
.PD
.SH "AUTHOR"
.IX Header "AUTHOR"
Clockwork was designed and written by James Hunt.


CHANGELOG
Release Notes for Clockwork


v2.3.0 - 2014-09-11

  * New `cw' multi-call binary to replace cw* tools into a single,
    user-extensible sub-command dispather, a la git.

    The following utilities have therefore been renamed:

       cwcc     ->  cw-cc
       fact     ->  cw-fact
       cwtrust  ->  cw-trust
       cwpol    ->  cw-shell
       cwkey    ->  cw-cert
       cw-run   ->  cw-mesh (and cw-show, cw-ping, etc.)

       cwtool   ->  cw-localsys [internal to cogd]

   * cw-mesh prompting is now done to stderr, for pipelines
   * cw-mesh now ignores nodes that optout by default
     (use --optout to retain the legacy behavior)
   * cw-mesh now honors ~/.cwrc as an adjunct to /etc/clockwork/cw.conf




res_file.c
#include <assert.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#include "res_file.h"
#include "pack.h"

#define RF_FD2FD_CHUNKSIZE 16384

#define RES_FILE_PACK_PREFIX "res_file::"
#define RES_FILE_PACK_OFFSET 10
/* Pack format for res_file structure:
     L - rf_enf
     a - rf_lpath
     a - rf_rsha1
     a - rf_owner
     a - rf_group
     L - rf_mode
 */
#define RES_FILE_PACK_FORMAT "LaaaaL"


static int _res_file_fd2fd(int dest, int src, ssize_t bytes);
static int _res_file_diff(struct res_file *rf);

/*****************************************************************/

static int _res_file_fd2fd(int dest, int src, ssize_t bytes)
{
	char buf[RF_FD2FD_CHUNKSIZE];
	ssize_t nread = 0;

	while (bytes > 0) {
		nread = (RF_FD2FD_CHUNKSIZE > bytes ? bytes : RF_FD2FD_CHUNKSIZE);

		if ((nread = read(src, buf, nread)) <= 0
		 || write(dest, buf, nread) != nread) {
			return -1;
		}

		bytes -= nread;
	}

	return 0;
}

/*
 * Calculate the difference between the expected file attributes
 * in a res_file structure, and the actual file attributes from
 * the filesystem.
 */
static int _res_file_diff(struct res_file *rf)
{
	assert(rf);

	rf->rf_diff = RES_FILE_NONE;

	if (res_file_enforced(rf, UID) && rf->rf_uid != rf->rf_stat.st_uid) {
		rf->rf_diff |= RES_FILE_UID;
	}
	if (res_file_enforced(rf, GID) && rf->rf_gid != rf->rf_stat.st_gid) {
		rf->rf_diff |= RES_FILE_GID;
	}
	if (res_file_enforced(rf, MODE) && (rf->rf_stat.st_mode & 07777) != rf->rf_mode) {
		rf->rf_diff |= RES_FILE_MODE;
	}
	if (res_file_enforced(rf, SHA1) && memcmp(rf->rf_rsha1.raw, rf->rf_lsha1.raw, SHA1_DIGEST_SIZE) != 0) {
		rf->rf_diff |= RES_FILE_SHA1;
	}

	return 0;
}

/*****************************************************************/

void* res_file_new(const char *key)
{
	struct res_file *rf;

	rf = xmalloc(sizeof(struct res_file));
	list_init(&rf->res);

	rf->rf_enf = 0;
	rf->rf_diff = 0;
	memset(&rf->rf_stat, 0, sizeof(struct stat));
	rf->rf_exists = 0;

	rf->rf_owner = NULL;
	rf->rf_uid = 0;

	rf->rf_group = NULL;
	rf->rf_gid = 0;

	rf->rf_mode = 0600; /* sane default... */

	rf->rf_lpath = NULL;
	rf->rf_rpath = NULL;

	sha1_init(&(rf->rf_lsha1), NULL);
	sha1_init(&(rf->rf_rsha1), NULL);

	if (key) {
		res_file_set(rf, "path", key);
		rf->key = string("res_file:%s", key);
	} else {
		rf->key = NULL;
	}

	return rf;
}

void res_file_free(void *res)
{
	struct res_file *rf = (struct res_file*)(res);
	if (rf) {
		list_del(&rf->res);

		free(rf->rf_rpath);
		free(rf->rf_lpath);

		free(rf->key);
	}

	free(rf);
}

const char* res_file_key(const void *res)
{
	const struct res_file *rf = (struct res_file*)(res);
	assert(rf);

	return rf->key;
}

int res_file_norm(void *res) {
	struct res_file *rf = (struct res_file*)(res);
	assert(rf);

	return sha1_file(rf->rf_rpath, &rf->rf_rsha1);
}

int res_file_set(void *res, const char *name, const char *value)
{
	struct res_file *rf = (struct res_file*)(res);
	assert(rf);

	if (strcmp(name, "owner") == 0) {
		free(rf->rf_owner);
		rf->rf_owner = strdup(value);
		rf->rf_enf |= RES_FILE_UID;

	} else if (strcmp(name, "group") == 0) {
		free(rf->rf_group);
		rf->rf_group = strdup(value);
		rf->rf_enf |= RES_FILE_GID;

	} else if (strcmp(name, "mode") == 0) {
		/* Mask off non-permission bits */
		rf->rf_mode = strtoll(value, NULL, 0) & 07777;
		rf->rf_enf |= RES_FILE_MODE;

	} else if (strcmp(name, "source") == 0) {
		free(rf->rf_rpath);
		rf->rf_rpath = strdup(value);
		if (sha1_file(rf->rf_rpath, &rf->rf_rsha1) != 0) {
			return -1;
		}
		rf->rf_enf |= RES_FILE_SHA1;

	} else if (strcmp(name, "path") == 0) {
		free(rf->rf_lpath);
		rf->rf_lpath = strdup(value);

	} else if (strcmp(name, "present") == 0) {
		if (strcmp(value, "no") != 0) {
			rf->rf_enf ^= RES_FILE_ABSENT;
		} else {
			rf->rf_enf |= RES_FILE_ABSENT;
		}

	} else {
		/* unknown attribute. */
		return -1;
	}

	return 0;
}

/*
 * Fill in the local details of res_file structure,
 * including invoking stat(2)
 */
int res_file_stat(void *res, const struct resource_env *env)
{
	struct res_file *rf = (struct res_file*)(res);

	assert(rf);
	assert(rf->rf_lpath);

	if (stat(rf->rf_lpath, &rf->rf_stat) == -1) { /* new file */
		rf->rf_diff = rf->rf_enf;
		rf->rf_exists = 0;
		return 0;
	}

	rf->rf_exists = 1;

	/* only generate sha1 checksums if necessary */
	if (res_file_enforced(rf, SHA1)) {
		if (sha1_file(rf->rf_lpath, &(rf->rf_lsha1)) == -1) {
			return -1;
		}
	}

	return _res_file_diff(rf);
}

struct report* res_file_fixup(void *res, int dryrun, const struct resource_env *env)
{
	struct res_file *rf = (struct res_file*)(res);
	assert(rf);
	assert(env);

	struct report *report = report_new("File", rf->rf_lpath);
	char *action;
	int new_file = 0;
	int local_fd = -1;

	/* Remove the file */
	if (res_file_enforced(rf, ABSENT)) {
		if (rf->rf_exists == 1) {
			action = string("remove file");

			if (dryrun) {
				report_action(report, action, ACTION_SKIPPED);
			} else if (unlink(rf->rf_lpath) == 0) {
				report_action(report, action, ACTION_SUCCEEDED);
			} else {
				report_action(report, action, ACTION_FAILED);
			}
		}

		return report;
	}

	if (!rf->rf_exists) {
		new_file = 1;
		action = string("create file");

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			local_fd = creat(rf->rf_lpath, rf->rf_mode);
			if (local_fd >= 0) {
				report_action(report, action, ACTION_SUCCEEDED);
			} else {
				report_action(report, action, ACTION_FAILED);
				return report;
			}
		}

		rf->rf_diff = rf->rf_enf;
		/* No need to chmod the file again */
		rf->rf_diff ^= RES_FILE_MODE;
	}

	if (res_file_different(rf, SHA1)) {
		assert(rf->rf_lpath);

		action = string("update content from master copy");
		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			if (local_fd < 0) {
				local_fd = open(rf->rf_lpath, O_CREAT | O_RDWR | O_TRUNC, rf->rf_mode);
				if (local_fd == -1) {
					report_action(report, action, ACTION_FAILED);
					return report;
				}
			}

			if (env->file_fd == -1) {
				report_action(report, action, ACTION_FAILED);
				return report;
			}

			if (_res_file_fd2fd(local_fd, env->file_fd, env->file_len) == -1) {
				report_action(report, action, ACTION_FAILED);
				return report;
			}

			report_action(report, action, ACTION_SUCCEEDED);
		}
	}

	if (res_file_different(rf, UID)) {
		if (new_file) {
			action = string("set owner to %s(%u)", rf->rf_owner, rf->rf_uid);
		} else {
			action = string("change owner from %u to %s(%u)", rf->rf_stat.st_uid, rf->rf_owner, rf->rf_uid);
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else if (chown(rf->rf_lpath, rf->rf_uid, -1) == 0) {
			report_action(report, action, ACTION_SUCCEEDED);
		} else {
			report_action(report, action, ACTION_FAILED);
		}
	}

	if (res_file_different(rf, GID)) {
		if (new_file) {
			action = string("set group to %s(%u)", rf->rf_group, rf->rf_gid);
		} else {
			action = string("change group from %u to %s(%u)", rf->rf_stat.st_gid, rf->rf_group, rf->rf_gid);
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else if (chown(rf->rf_lpath, -1, rf->rf_gid) == 0) {
			report_action(report, action, ACTION_SUCCEEDED);
		} else {
			report_action(report, action, ACTION_FAILED);
		}
	}

	if (res_file_different(rf, MODE)) {
		if (new_file) {
			action = string("set permissions to %04o", rf->rf_mode);
		} else {
			action = string("change permissions from %04o to %04o", rf->rf_stat.st_mode & 07777, rf->rf_mode);
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else if (chmod(rf->rf_lpath, rf->rf_mode) == 0) {
			report_action(report, action, ACTION_SUCCEEDED);
		} else {
			report_action(report, action, ACTION_FAILED);
		}
	}

	return report;
}

int res_file_is_pack(const char *packed)
{
	assert(packed);
	return strncmp(packed, RES_FILE_PACK_PREFIX, RES_FILE_PACK_OFFSET);
}

char* res_file_pack(const void *res)
{
	const struct res_file *rf = (const struct res_file*)(res);
	assert(rf);

	return pack(RES_FILE_PACK_PREFIX, RES_FILE_PACK_FORMAT,
	            rf->rf_enf,
	            rf->rf_lpath, rf->rf_rsha1.hex, rf->rf_owner, rf->rf_group, rf->rf_mode);
}

void* res_file_unpack(const char *packed)
{
	char *hex = NULL;
	struct res_file *rf = res_file_new(NULL);

	if (unpack(packed, RES_FILE_PACK_PREFIX, RES_FILE_PACK_FORMAT,
		&rf->rf_enf,
		&rf->rf_lpath, &hex, &rf->rf_owner, &rf->rf_group, &rf->rf_mode) != 0) {

		free(hex);
		res_file_free(rf);
		return NULL;
	}

	sha1_init(&rf->rf_rsha1, hex);
	free(hex);

	rf->key = string("res_file:%s", rf->rf_lpath);

	return rf;
}

